File: java/example_code/s3/src/main/java/aws/example/s3/SetAcl.java
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
package aws.example.s3;

import com.amazonaws.AmazonServiceException;
import com.amazonaws.regions.Regions;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.model.AccessControlList;
import com.amazonaws.services.s3.model.EmailAddressGrantee;
import com.amazonaws.services.s3.model.Permission;

/**
 * Add a bucket policy to an existing S3 bucket.
 *
 * This code expects that you have AWS credentials set up per:
 * http://docs.aws.amazon.com/java-sdk/latest/developer-guide/setup-credentials.html
 */
public class SetAcl {
    public static void setBucketAcl(String bucket_name, String email, String access) {
        System.out.format("Setting %s access for %s\n", access, email);
        System.out.println("on bucket: " + bucket_name);

        final AmazonS3 s3 = AmazonS3ClientBuilder.standard().withRegion(Regions.DEFAULT_REGION).build();
        try {
            // get the current ACL
            AccessControlList acl = s3.getBucketAcl(bucket_name);
            // set access for the grantee
            EmailAddressGrantee grantee = new EmailAddressGrantee(email);
            Permission permission = Permission.valueOf(access);
            acl.grantPermission(grantee, permission);
            s3.setBucketAcl(bucket_name, acl);
        } catch (AmazonServiceException e) {
            System.err.println(e.getErrorMessage());
            System.exit(1);
        }
    }

    public static void setObjectAcl(String bucket_name, String object_key, String email, String access) {
        System.out.format("Setting %s access for %s\n", access, email);
        System.out.println("for object: " + object_key);
        System.out.println(" in bucket: " + bucket_name);

        final AmazonS3 s3 = AmazonS3ClientBuilder.standard().withRegion(Regions.DEFAULT_REGION).build();
        try {
            // get the current ACL
            AccessControlList acl = s3.getObjectAcl(bucket_name, object_key);
            // set access for the grantee
            EmailAddressGrantee grantee = new EmailAddressGrantee(email);
            Permission permission = Permission.valueOf(access);
            acl.grantPermission(grantee, permission);
            s3.setObjectAcl(bucket_name, object_key, acl);
        } catch (AmazonServiceException e) {
            System.err.println(e.getErrorMessage());
            System.exit(1);
        }
    }

    public static void main(String[] args) {
        final String USAGE = "\n" +
                "Usage:\n" +
                "  SetAcl <bucket> [object] <email> <permission>\n\n" +
                "Where:\n" +
                "  bucket     - the bucket to grant permissions on\n" +
                "  object     - (optional) the object grant permissions on\n" +
                "               If object is specified, granted permissions will be\n" +
                "               for the object, not the bucket.\n" +
                "  email      - The email of the user to set permissions for\n" +
                "  permission - The permission(s) to set. Can be one of:\n" +
                "               FullControl, Read, Write, ReadAcp, WriteAcp\n\n" +
                "Examples:\n" +
                "    SetAcl testbucket user@example.com read\n" +
                "    SetAcl testbucket testobject user@example.com write\n\n";

        if (args.length < 3) {
            System.out.println(USAGE);
            System.exit(1);
        }

        int cur_arg = 0;

        String bucket_name = args[cur_arg++];
        String object_key = (args.length > 3) ? args[cur_arg++] : null;
        String email = args[cur_arg++];
        String access = args[cur_arg++];

        if (object_key != null) {
            setObjectAcl(bucket_name, object_key, email, access);
        } else {
            setBucketAcl(bucket_name, email, access);
        }

        System.out.println("Done!");
    }
}

File: .kotlin_alpha/services/s3/src/main/kotlin/com/kotlin/s3/SetAcl.kt
//snippet-sourcedescription:[SetAcl.kt demonstrates how to set a new access control list (ACL) for an Amazon Simple Storage Service (Amazon S3) bucket.]
//snippet-keyword:[AWS SDK for Kotlin]
//snippet-keyword:[Code Sample]
//snippet-service:[Amazon S3]
//snippet-sourcetype:[full-example]
//snippet-sourcedate:[11/05/2021]
//snippet-sourceauthor:[scmacdon-aws]

/*
   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
   SPDX-License-Identifier: Apache-2.0
*/

package com.kotlin.s3

// snippet-start:[s3.kotlin.set_acl.import]
import aws.sdk.kotlin.services.s3.S3Client
import aws.sdk.kotlin.services.s3.model.Grantee
import aws.sdk.kotlin.services.s3.model.Type
import aws.sdk.kotlin.services.s3.model.Grant
import aws.sdk.kotlin.services.s3.model.Permission
import aws.sdk.kotlin.services.s3.model.Owner
import aws.sdk.kotlin.services.s3.model.AccessControlPolicy
import aws.sdk.kotlin.services.s3.model.PutBucketAclRequest
import aws.sdk.kotlin.services.s3.model.S3Exception
import kotlin.system.exitProcess
// snippet-end:[s3.kotlin.set_acl.import]

/**
To run this Kotlin code example, ensure that you have setup your development environment,
including your credentials.

For information, see this documentation topic:
https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/setup.html
 */
suspend fun main(args: Array<String>) {

    val usage = """
    Usage:
        <bucketName> <id>

    Where:
        bucketName - the Amazon S3 bucket to grant permissions on.
        id - the ID of the owner of this bucket (you can get this value from the AWS Management Console under the Amazon S3 Access control list (ACL)).
    """

     if (args.size != 2) {
         println(usage)
         exitProcess(0)
     }

    val bucketName = args[0]
    val id = args[1]
    val s3Client = S3Client { region = "us-east-1" }
    setBucketAcl(s3Client, bucketName, id)
    s3Client.close()
}

// snippet-start:[s3.kotlin.set_acl.main]
suspend fun setBucketAcl(s3Client: S3Client, bucketName: String, idVal: String) {

   try {
        val myGrant = Grantee {
             id = idVal
             type = Type.CanonicalUser
        }

        val ownerGrant = Grant {
            grantee = myGrant
            permission = Permission.FullControl
         }

        val grantList2 = mutableListOf<Grant>()
        grantList2.add(ownerGrant)

        val ownerOb = Owner {
           id = idVal
         }

        val acl = AccessControlPolicy {
            owner = ownerOb
            grants = grantList2
        }

        val putAclReq = PutBucketAclRequest {
            bucket = bucketName
            accessControlPolicy = acl
        }

       s3Client.putBucketAcl(putAclReq)
       println("An ACL was successfully set on $bucketName")

   } catch (e: S3Exception) {
       println(e.message)
       s3Client.close()
       exitProcess(0)
   }
}
// snippet-end:[s3.kotlin.set_acl.main]